feat: `gt_data_color_by_group()`

avamsi opened the pull request:
closes #83

Change-Id: Id000000048a4ee276613f35c991b6e934afdbd89

avamsi commented:
@juleswg23 PTAL! The code works for me locally but I couldn't get the snapshots to render colors -- any idea why?

juleswg23 commented on gt_extras/tests/test_colors.py:
```diff
@@ -1,11 +1,33 @@
 
+def test_gt_data_color_by_group_no_group(snapshot):
+    df = pd.DataFrame({"B": [1, 2, 3, 4, 5]})
+    assert_rendered_body(snapshot, gt_data_color_by_group(GT(df)))
```
Using this function, because of how we grab the groups, will require that the user set the `groupname_col`.

```python
df = pd.DataFrame({
    "group": ["A", "B", "C", "D", "E"], 
    "values": [1, 2, 3, 4, 5]
})
gt = GT(df, groupname_col="group")
gt_data_color_by_group(gt, columns="values")
```

avamsi replied:
This was intentional -- note the test name ending with `no_group`. Other tests do have a groupname_col arg as you'd expect.

avamsi commented on gt_extras/colors.py:
```diff
@@ -15,7 +15,40 @@
+    """
+    for group in gt._stub.group_rows:
+        gt = gt.data_color(columns, group.indices, palette)
+    return gt
```
I don't think this is needed:

```python
def gt_data_color_by_group(gt, columns=None, palette=None) -> GT:
    for group in gt._stub.group_rows:
        print(type(group.indices), list(map(type, group.indices)), group.indices)
        gt = gt.data_color(columns, group.indices, palette)
    return gt


df = pl.DataFrame({"a": [1, 2, 2, 3, 3, 3], "b": range(6)})
GT(df, groupname_col="a").pipe(gt_data_color_by_group, palette="GnBu")
```

This prints:

```
<class 'list'> [<class 'int'>] [0]
<class 'list'> [<class 'int'>, <class 'int'>] [1, 2]
<class 'list'> [<class 'int'>, <class 'int'>, <class 'int'>] [3, 4, 5]
```

So, `group.indices` is `list[int]` already and works with `GT.data_color`.

avamsi replied:
Ah, sorry -- good catch, thank you!

avamsi reviewed (commented):
The following code works fine for me in a notebook (with per group colors like I'd expect) -- but somehow in the gt-extras tests, the snapshots I'm generating don't have any colors (I don't know if I'm missing any dev dependencies or something), could you help fix that?

```python
def gt_data_color_by_group(gt, columns=None, palette=None) -> GT:
    for group in gt._stub.group_rows:
        print(type(group.indices), list(map(type, group.indices)), group.indices)
        gt = gt.data_color(columns, group.indices, palette)
    return gt


df = pl.DataFrame({"a": [1, 2, 2, 3, 3, 3], "b": range(6)})
GT(df, groupname_col="a").pipe(gt_data_color_by_group, palette="GnBu")
```

avamsi commented:
Fixed the bug and added tests for Pandas and Polars separately to make sure both are passing. Also, split the test for multiple groups into singletons and all groups, to distinguish the palette. PTAL, thanks!

juleswg23 commented on gt_extras/colors.py:
```diff
@@ -15,7 +15,40 @@
+        access from ColorBrewer* section). If `None`, then a default palette will be used.
+    """
+    for group in gt._stub.group_rows:
+        gt = gt.data_color(columns, list(map(int, group.indices)), palette)
```
To give the user has full control over data_color, could we do this?

```python
def gt_data_color_by_group(gt: GT, columns: SelectExpr = None, **kwargs) -> GT:
    for group in gt._stub.group_rows:
        gt = gt.data_color(columns, list(map(int, group.indices)), **kwargs)
    return gt
```

If we do that we'd have to update the docstring appropriately.

```
**kwargs
    Additional arguments passed to [`GT.data_color`](https://posit-dev.github.io/great-tables/reference/GT.data_color.html#great_tables.GT.data_color).
    This includes `palette`, `domain`, `na_color`, `alpha`, `reverse`, `autocolor_text`,
    and `truncate`.
```

juleswg23 commented on gt_extras/tests/test_colors.py:
```diff
@@ -1,11 +1,63 @@
+def test_gt_data_color_by_group_pl_multiple_groups(snapshot):
+    df = pl.DataFrame({"A": [1, 2, 2, 3, 3, 3], "B": [1, 2, 3, 4, 5, 6]})
+    gt = gt_data_color_by_group(GT(df, groupname_col="A"))
+    assert_rendered_body(snapshot, gt)
```
These extra tests look great! It's a little picky, but I'd hope to have slightly fewer snapshot tests. Maybe one for pandas and one for polars, and the rest could follow this form?

```python
def test_gt_data_color_by_group_pd_no_groups():
    df = pd.DataFrame({"B": [1, 2, 3, 4, 5, 6]})
    gt = GT(df)
    html = gt_data_color_by_group(gt).as_raw_html()

    assert 'style="color:' not in html


def test_gt_data_color_by_group_pd_single_group():
    df = pd.DataFrame({"A": [1, 1, 1, 1, 1, 1], "B": [1, 2, 3, 4, 5, 6]})
    gt = GT(df, groupname_col="A")
    html = gt_data_color_by_group(gt).as_raw_html()

    assert "color: #FFFFFF; background-color: #000000;" in html
    assert "color: #000000; background-color: #ad8560;" in html
    assert "color: #000000; background-color: #2fa2c8;" in html


def test_gt_data_color_by_group_pd_multiple_singletons():
    df = pd.DataFrame({"A": [1, 2, 3, 4, 5, 6], "B": [1, 2, 3, 4, 5, 6]})
    gt = GT(df, groupname_col="A")
    html = gt_data_color_by_group(gt).as_raw_html()

    assert html.count("color: #FFFFFF; background-color: #000000;") == 6
```

juleswg23 commented on gt_extras/tests/test_colors.py:
```diff
@@ -1,17 +1,73 @@
+def test_gt_data_color_by_group_pd_multiple_groups(snapshot):
+    df = pd.DataFrame({"A": [1, 2, 2, 3, 3, 3], "B": [1, 2, 3, 4, 5, 6]})
+    gt = gt_data_color_by_group(GT(df, groupname_col="A"))
+    assert_rendered_body(snapshot, gt)
```
Sorry to be a bother with tests – I didn't notice this before but it's nagging at me that we have essentially the same four tests duplicated twice. For snapshot tests, we can reference the same snapshot like this:

```python
def test_gt_data_color_by_group_multiple_groups(snapshot):
    data = {"A": [1, 2, 2, 3, 3, 3], "B": [1, 2, 3, 4, 5, 6]}
    for df in [pd.DataFrame(data), pl.DataFrame(data)]:
        gt = gt_data_color_by_group(GT(df, groupname_col="A"))
        assert_rendered_body(snapshot(name="pl_and_pd"), gt)
```

There are probably other ways to combine the snapshot, this is just the first I came up with.

avamsi replied:
No worries, updated -- thanks!

juleswg23 commented:
Looks great, thank you for all your work on this!
